sql/plan/union.go

```diff
@@ -0,0 +1,88 @@
+package plan
+
+import (
+	"io"
+
+	"github.com/src-d/go-mysql-server/sql"
+)
+
+// Union is a node that returns everything in Left and then everything in Right
+type Union struct {
+	BinaryNode
+}
+
+// NewUnion creates a new Union node with the given children.
+func NewUnion(left, right sql.Node) *Union {
+	return &Union{
+		BinaryNode: BinaryNode{Left: left, Right: right},
+	}
+}
+
+func (u *Union) Schema() sql.Schema {
+	return u.Left.Schema()
+}
+
+// RowIter implements the Node interface.
+func (u *Union) RowIter(ctx *sql.Context) (sql.RowIter, error) {
+	span, ctx := ctx.Span("plan.Union")
+	li, err := u.Left.RowIter(ctx)
+	if err != nil {
+		span.Finish()
+		return nil, err
+	}
+	ui := &unionIter{
+		li,
+		func() (sql.RowIter, error) {
+			return u.Right.RowIter(ctx)
+		},
+	}
+	return sql.NewSpanIter(span, ui), nil
+}
+
+// WithChildren implements the Node interface.
+func (u *Union) WithChildren(children ...sql.Node) (sql.Node, error) {
+	if len(children) != 2 {
+		return nil, sql.ErrInvalidChildrenNumber.New(u, len(children), 2)
+	}
+	return NewUnion(children[0], children[1]), nil
+}
+
+func (u Union) String() string {
+	pr := sql.NewTreePrinter()
+	_ = pr.WriteNode("Union")
+	_ = pr.WriteChildren(u.Left.String(), u.Right.String())
+	return pr.String()
+}
+
+type unionIter struct {
+	cur      sql.RowIter
+	nextIter func() (sql.RowIter, error)
+}
+
+func (ui *unionIter) Next() (sql.Row, error) {
+	res, err := ui.cur.Next()
+	if err == io.EOF {
+		if ui.nextIter == nil {
+			return nil, io.EOF
+		}
+		err = ui.cur.Close()
+		if err != nil {
+			return nil, err
+		}
+		ui.cur, err = ui.nextIter()
+		ui.nextIter = nil
+		if err != nil {
+			return nil, err
+		}
+		return ui.cur.Next()
```

Comment:
Probably should recurse here instead

Reply:
I think it's (much) clearer without recursion. Am I crazy?

Reply:
The issue is that this is only correct when N=2. For the general case of composing an iterator from N sub iterators, if ui.cur.Next() on line 87 returns (nil, io.EOF), then recursion correctly continues to the next element, but what you have now terminates early.

Reply:
But that's not what this method does...agreed if this iterator took a bunch of row iterators and concat'd them, this impl would need to be different.

Reply:
I still think recursion is cleaner, but you do you :p